### PR TITLE
Hash op

### PR DESCRIPTION
**Pull request by Meiyim**
Op implemention of XXHash

### CODE REVIEW

**Review by reyoung — COMMENTED** [review]
You should change `CMakeLists.txt` in `operators` directory to make `hash_op` depends on `xxhash`.

The sample code is here

https://github.com/PaddlePaddle/Paddle/blob/a615ad46e4f8a232deae9db7558e5277f057e208/paddle/fluid/operators/CMakeLists.txt#L268-L277

File: cmake/external/xxhash.cmake [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+set(XXHASH_INSTALL_DIR ${THIRD_PARTY_PATH}/install/xxhash)
+set(XXHASH_INCLUDE_DIR "${XXHASH_INSTALL_DIR}/include")
+
+
+ExternalProject_Add(
+    extern_xxhash
+    ${EXTERNAL_PROJECT_LOG_ARGS}
+    GIT_REPOSITORY  "https://github.com/Cyan4973/xxHash"
+    # eigen on cuda9.1 missing header of math_funtions.hpp
+    # https://stackoverflow.com/questions/43113508/math-functions-hpp-not-found-when-using-cuda-with-eigen
```
Comment (reyoung) [inline cmake/external/xxhash.cmake:13]: Why this line？

File: cmake/external/xxhash.cmake [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+
+set(XXHASH_LIBRARIES "${XXHASH_INSTALL_DIR}/lib/libxxhash.a")
+INCLUDE_DIRECTORIES(${XXHASH_INCLUDE_DIR})
+
+add_library(xxhash STATIC IMPORTED GLOBAL)
+set_property(TARGET xxhash PROPERTY IMPORTED_LOCATION ${XXHASH_LIBRARIES})
+#if (${CMAKE_VERSION} VERSION_LESS "3.3.0")
+#    set(dummyfile ${CMAKE_CURRENT_BINARY_DIR}/lib_xxhash_dummy.c)
+#    file(WRITE ${dummyfile} "const char * dummy_any = \"${dummyfile}\";")
```
Comment (reyoung) [inline cmake/external/xxhash.cmake:34]: either uncomment these lines or remove these lines

File: python/paddle/fluid/tests/unittests/test_hash_op.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+class TestScaleOp(OpTest):
+    def setUp(self):
+        self.op_type = "hash"
+        self.init_test_case()
+        self.inputs = {'X': (self.in_seq, self.lod)}
+        self.attrs = {'num_hash': 8, 'mod_by': 10000}
+        self.outputs = {'Out': (self.out_seq, self.lod)}
+
+    def init_test_case(self):
```
Comment (kuke) [inline python/paddle/fluid/tests/unittests/test_hash_op.py:28]: 这里应该可以写个有效的单测 https://pypi.org/project/xxhash/

**Comment by Meiyim** [comment]
link error after adding `op_library(xxhash_op DEPS xxhash) `:
```shell
/usr/bin/ld: ../../../third_party/install/xxhash/lib/libxxhash.a(xxhash.o): relocation R_X86_64_32S against `.rodata' can not be used when making a shared object; recompile with -fPIC
```

libxxhash.a is good if link with a test cpp: g++ test.cpp -lxxhash
.a is the only output of xxhash dependency
```shell
# ls third_party/install/xxhash/lib/
libxxhash.a
```

BTW, xxhash is a C(not C++) code. 

any ideas? @reyoung 
thanks ~

**Comment by velconia** [comment]
I will take a look at this

**Comment by velconia** [comment]
@Meiyim thanks for your work, and the reason of your problems is that: the ```libxxhash``` should be compiled with ```shared mode```, so we should add ```-fPIC (Position Independent Code)``` to the compilation flags when compiling ```libxxhash```, there are two ways to add the flags:

1. use ```sed``` to change the ```Makefile``` of ```libxxhash```

in the file ```cmake/external/xxhash.cmake```, change ```BUILD_COMMAND``` to 

```
BUILD_COMMAND     sed -i "s/-Wstrict-prototypes -Wundef/-Wstrict-prototypes -Wundef -fPIC/g" ${XXHASH_SOURCE_DIR}/src/extern_xxhash/Makefile && make lib
```

2. I forked the ```xxhash``` repo [here](https://github.com/velconia/xxHash/tree/paddle/v0.6.5), and I add the ```-fPIC``` flags to the Makefile, you could use the 'local repo' too. just change your 
```cmake/external/xxhash.cmake``` file to 

```
GIT_REPOSITORY  "https://github.com/velconia/xxHash"
GIT_TAG         "paddle/v0.6.5"
```

you could use either way listed above to solve your problem.

File: python/paddle/fluid/tests/unittests/test_hash_op.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        self.in_seq = np.random.randint(0, 10, (30, 1)).astype("int32")
```
Comment (velconia) [inline python/paddle/fluid/tests/unittests/test_hash_op.py:29]: np.random是否需要固定seed

**Pull request closed without merging**